power_grid_model_c/power_grid_model/include/power_grid_model/main_core/update.hpp

```diff
@@ -130,4 +135,194 @@ inline void update_inverse(MainModelState<ComponentContainer> const& state, Forw
+                                                     });
+                                                 }) &&
+                             !properties.ids_all_na;
+
+    if (all_spans.empty()) {
+        properties.update_ids_match = true;
+    } else {
+        // Remember the begin iterator of the first scenario, then loop over the remaining scenarios and
+        // check the ids
+        auto const first_span = all_spans[0];
+        // check the subsequent scenarios
+        // only return true if all scenarios match the ids of the first batch
+        properties.update_ids_match =
+            std::ranges::all_of(all_spans.cbegin() + 1, all_spans.cend(), [&first_span](auto const& current_span) {
+                return std::ranges::equal(
+                    current_span, first_span,
+                    [](typename CompType::UpdateType const& obj, typename CompType::UpdateType const& first) {
+                        return obj.id == first.id;
+                    });
+            });
+    }
+}
+
+template <class CompType>
+detail::UpdateCompProperties check_component_independence(ConstDataset const& update_data, Idx n_component) {
+    detail::UpdateCompProperties properties;
+    properties.name = CompType::name;
+    auto const component_idx = update_data.find_component(properties.name, false);
+    properties.is_columnar = update_data.is_columnar(properties.name);
+    properties.dense = update_data.is_dense(properties.name);
+    properties.uniform = update_data.is_uniform(properties.name);
+    properties.has_any_elements =
+        component_idx != invalid_index && update_data.get_component_info(component_idx).total_elements > 0;
+    properties.elements_ps_in_update =
+        properties.uniform ? update_data.uniform_elements_per_scenario(properties.name) : invalid_index;
+    properties.elements_in_base = n_component;
+
+    if (properties.is_columnar) {
+        detail::process_buffer_span<CompType>(
+            update_data.template get_columnar_buffer_span_all_scenarios<meta_data::update_getter_s, CompType>(),
+            properties);
+    } else {
+        detail::process_buffer_span<CompType>(
+            update_data.template get_buffer_span_all_scenarios<meta_data::update_getter_s, CompType>(), properties);
+    }
+
+    return properties;
+}
+
+inline void validate_update_data_independence(detail::UpdateCompProperties const& comp) {
+    if (comp.is_empty_component()) {
+        return; // empty dataset is still supported
+    }
+    auto const elements_ps = comp.get_n_elements();
+    assert(comp.uniform || elements_ps < 0);
+
+    if (elements_ps >= 0 && comp.elements_in_base < elements_ps) {
+        throw DatasetError("Update data has more elements per scenario than input data for component " + comp.name +
+                           "!");
+    }
+    if (comp.ids_part_na) {
+        throw DatasetError("Some IDs are not valid for component " + comp.name + " in update data!");
```

Comment:
Now that I re-read this, `valid` can be vague here. I would argue that we go back towards the old one, clear albeit mouthful.
```suggestion
        throw DatasetError("IDs contain both numbers and NANs for component " + comp.name + " in update data!");
```

Reply:
I actually like your suggestion better. More verbose, but clearer.